File: 79-word-search/word-search.cpp
class Solution {
public:
    bool check(vector<vector<char>>& board, string word, int& count, int i, int j ){
        if(count == 0) return true;
        bool ans;
        if(i>=board.size() || j>=board[i].size() || i<0 || j<0 || word[count-1]!=board[i][j]){
            return false;
        }else{
        char temp = board[i][j];
        board[i][j] = '%';
        count--;
        ans = check(board, word, count,i+1,j)|| check(board, word, count,i,j+1)||check(board,word,count,i-1,j)||check(board, word, count,i,j-1);
        count++;
        board[i][j] = temp;
        }
        return ans;
    }
    bool exist(vector<vector<char>>& board, string word) {
        int count = word.size();
        for(int i=0;i<board.size();i++){
            for(int j=0;j<board[i].size();j++){
                if(board[i][j]==word[count-1] && check(board,word,count,i,j)==true) return true;
            }
        }
        return false;
    }
};

File: 79-word-search/word-search.java
class Solution {
    private int[][] dir= {{0,1},{-1,0},{1,0},{0,-1}};

    public boolean exist(char[][] board, String word) {
        int index = 0;

        for(int i=0;i<board.length;i++){
            for(int j=0;j<board[0].length;j++){
                  if(Character.compare(board[i][j],word.charAt(index))==0){
                      if(existHelper(i,j,board,word,index)==true) return true;
                  }
            }
        }
        return false;
    }
    private boolean existHelper(int r,int c,char[][] board, String word,int in){
    if(in==word.length()){
        return true;
    }
    if(r>-1 && r<board.length && c>-1 && c<board[0].length) 
    {
        if(Character.compare(board[r][c],word.charAt(in))!=0) return false;
        
    }
    else return false;
    char temp = board[r][c];
    board[r][c]='#';
    for(int i=0;i<4;i++){
        int rr = r+dir[i][0];
        int cc = c+dir[i][1];
        if(existHelper(rr,cc,board,word,in+1)) return true;
    }
    board[r][c]=temp;

    return false;
}
}